lib/virtio_mmio/virtio_rng.c

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2022 Wind River Systems, Inc.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * $FreeBSD$
+ */
+
+#include <openamp/open_amp.h>
+#include <openamp/virtqueue.h>
+#include <openamp/virtio.h>
+#include <openamp/virtio_mmio.h>
+#include <metal/device.h>
+
+int virtio_mmio_rng_init(struct virtio_device *vdev, struct virtqueue **vqs,
+			 char **vq_names, int vq_count)
+{
+	uint32_t devid, features;
+	struct virtqueue *vq = NULL;
+	int i;
+
+	if (!vdev || !vqs || !vq_names) {
+		return -EINVAL;
+	}
+
+	vdev->vrings_info = metal_allocate_memory(sizeof(struct virtio_vring_info) * vq_count);
+	if (!vdev->vrings_info) {
+		return -ENOMEM;
+	}
+
+	devid = virtio_get_devid(vdev);
+	if (devid != VIRTIO_ID_ENTROPY) {
+		metal_log(METAL_LOG_ERROR, "Expected devid %04x, got %04x\n",
+			  VIRTIO_ID_ENTROPY, devid);
+		return -ENODEV;
```

Comment:
vdev->vrings_info should be free in case of error in the whole function

Reply:
Done.